File: CpuA32/assembler.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA32/assembler.h"
#include "CpuA32/opcode_gen.h"
#include "CpuA32/symbolic.h"
#include "Util/parse.h"

#include <map>

namespace cwerg::a32 {
using namespace cwerg::elf;

// +-prefix converts an enum the underlying type
template <typename T>
constexpr auto operator+(T e) noexcept
    -> std::enable_if_t<std::is_enum<T>::value, std::underlying_type_t<T>> {
  return static_cast<std::underlying_type_t<T>>(e);
}

std::string_view padding_four_zero("\0\0\0\0", 4);
std::string_view padding_zero("\0", 1);
std::string_view padding_nop("\x00\xf0\x20\xe3", 4);

const char ARM_ATTRIBUTES[] = {0x41, 0x11, 0, 0, 0, 0x61, 0x65, 0x61, 0x62,
                               0x69, 0,    1, 7, 0, 0,    0,    8,    1};

void Unit::AddIns(Ins* ins) {
  if (ins->has_reloc()) {
    const elf::Symbol<uint32_t>* sym =
        FindOrAddSymbol(ins->reloc_symbol, ins->is_local_sym);
    AddReloc(+ins->reloc_kind, sec_text, sym, ins->operands[ins->reloc_pos]);
    ins->clear_reloc();
  }
  uint32_t data = Assemble(*ins);
  sec_text->AddData({(const char*)&data, 4});
}

std::array<std::string_view, 6> kStartupCode = {  //
    "ldr_imm_add al r0 sp 0",                     //
    "add_imm al r1 sp 4",                         //
    "bl expr:call:main",                       //
    "movw al r7 1",                               //
    "svc al 0",                                   //
    "ud2 al"};

void Unit::AddStartupCode() {
  FunStart("_start", 16, padding_nop);
  std::vector<std::string_view> token;
  for (const auto line : kStartupCode) {
    token.clear();
    if (!ParseLineWithStrings(line, false, &token)) {
      ASSERT(false, "bad internal code template " << line);
    }
    Ins ins;
    if (!InsFromSymbolized(token, &ins)) {
      ASSERT(false, "internal parse error " << token[0]);
    }
    AddIns(&ins);
  }
  FunEnd();
}

void Unit::AddLinkerDefs() {
  if (sec_bss->shdr.sh_size > 0) {
    sec_bss->PadData(16, padding_zero);
    AddSymbol("$$rw_data_end", sec_bss, false);
  } else if (sec_data->shdr.sh_size > 0) {
    sec_data->PadData(16, padding_zero);
    AddSymbol("$$rw_data_end", sec_data, false);
  }
}

bool HandleDirective(Unit* unit, const std::vector<std::string_view>& token) {
  const auto mnemonic = token[0];
  if (mnemonic == ".fun") {
    ASSERT(token.size() == 3, "");
    unit->FunStart(token[1], ParseInt<uint32_t>(token[2]).value(), padding_nop);
  } else if (mnemonic == ".endfun") {
    unit->FunEnd();
  } else if (mnemonic == ".mem") {
    ASSERT(token.size() == 4, "");
    unit->MemStart(token[1], ParseInt<uint32_t>(token[2]).value(), token[3],
                   padding_zero, false);
  } else if (mnemonic == ".localmem") {
    ASSERT(token.size() == 4, "");
    unit->MemStart(token[1], ParseInt<uint32_t>(token[2]).value(), token[3],
                   padding_zero, true);
  } else if (mnemonic == ".endmem") {
    unit->MemEnd();
  } else if (mnemonic == ".data") {
    ASSERT(token.size() == 3, "");
    size_t len = token[2].size();
    if (token[2][0] != '"' || token[2][len - 1] != '"' || len <= 2) {
      std::cerr << "malformed .data directive\n";
      return false;
    }
    char buffer[1024];
    len = EscapedStringToBytes({token[2].data() + 1, len - 2}, buffer);
    if (len == 0) {
      std::cerr << "malformed .data directive cannot parse payload\n";
      return false;
    }
    unit->AddData(ParseInt<uint32_t>(token[1]).value(), buffer, len);
  } else if (mnemonic == ".addr.mem") {
    ASSERT(token.size() == 4, "");
    unit->AddMemAddr(ParseInt<uint32_t>(token[1]).value(), +RELOC_TYPE_ARM::ABS32, token[2],
                     ParseInt<uint32_t>(token[3]).value());
  } else if (mnemonic == ".addr.fun") {
    ASSERT(token.size() == 3, "");
    unit->AddFunAddr(ParseInt<uint32_t>(token[1]).value(), +RELOC_TYPE_ARM::ABS32, token[2]);
  } else if (mnemonic == ".addr.bbl") {
    ASSERT(token.size() == 3, "");
    unit->AddBblAddr(ParseInt<uint32_t>(token[1]).value(), +RELOC_TYPE_ARM::ABS32, token[2]);
  } else if (mnemonic == ".bbl") {
    ASSERT(token.size() == 3, "");
    unit->AddLabel(token[1], ParseInt<uint32_t>(token[2]).value(), padding_nop);
  } else {
    std::cerr << "unknown directive " << mnemonic << "\n";
    return false;
  }
  return true;
}




bool HandleOpcode(Unit* unit, const std::vector<std::string_view>& token) {
  Ins ins;
  if (!InsFromSymbolized(token, &ins)) {
    std::cerr << "parse error " << token[0] << "\n";
    return false;
  }
  unit->AddIns(&ins);
  return true;
}

Unit::Unit()
    : sec_rodata(new Section<uint32_t>()),
      sec_text(new Section<uint32_t>()),
      sec_data(new Section<uint32_t>()),
      sec_bss(new Section<uint32_t>()) {
  sec_rodata->InitRodata(1);

  sec_text->InitText(1);

  sec_data->InitData(1);

  sec_bss->InitBss(1);
}

bool UnitParse(std::istream* input, bool add_startup_code, Unit* unit) {
  Unit out;
  int line_num = 0;
  std::vector<std::string_view> token;
  for (std::string line; getline(*input, line);) {
    ++line_num;
    token.clear();
    if (!ParseLineWithStrings(line.c_str(), false, &token)) {
      std::cerr << "Error while parsing line " << line_num << "\n" << line;
      return false;
    }

    // std::cerr << "@@reading: " << line << "\n";
    if (token.empty()) continue;
    if (token[0][0] == '#') continue;
    if (token.back()[0] == '#') token.pop_back();  // get rid of comment
    // Directives
    if (token[0][0] == '.') {
      if (!HandleDirective(unit, token)) return false;
    } else {
      if (!HandleOpcode(unit, token)) return false;
    }
  }
  unit->AddLinkerDefs();
  if (add_startup_code) {
    unit->AddStartupCode();
  }
  return true;
}



int32_t BranchOffset(const Reloc<uint32_t>& rel, int32_t sym_val) {
  return int32_t(sym_val - rel.section->shdr.sh_addr - rel.rel.r_offset - 8) >>
         2;
}

void ApplyRelocation(const Reloc<uint32_t>& rel) {
  void* patch_addr = (char*)rel.section->data->data() + rel.rel.r_offset;
  const uint32_t old_data = *(uint32_t*)patch_addr;
  const int32_t sym_val = rel.symbol->sym.st_value + rel.rel.r_addend;

  uint32_t new_data;
  switch (RELOC_TYPE_ARM(rel.rel.r_type)) {
    case RELOC_TYPE_ARM::ABS32:
      new_data = sym_val;
      break;
    case RELOC_TYPE_ARM::JUMP24:
      new_data = PatchIns(old_data, 1, BranchOffset(rel, sym_val));
      break;
    case RELOC_TYPE_ARM::CALL:
      new_data = PatchIns(old_data, 1, BranchOffset(rel, sym_val));
      break;
    case RELOC_TYPE_ARM::MOVW_ABS_NC:
      new_data = PatchIns(old_data, 2, sym_val & 0xffff);
      break;
    case RELOC_TYPE_ARM::MOVT_ABS:
      new_data = PatchIns(old_data, 2, (sym_val >> 16) & 0xffff);
      break;
    default:
      ASSERT(false, "unknown relocation type " << rel.rel.r_type);
      return;
  }
#if 0
  std::cout << "PATCH INS " << rel.rel.r_type << " " << std::hex
            << rel.rel.r_offset << " " << sym_val << " " << old_data << " "
            << new_data << std::dec << " " << rel.symbol->name << "\n";
#endif
  *(uint32_t*)patch_addr = new_data;
}

Executable<uint32_t> Unit::MakeExe(bool create_sym_tab) {
  std::vector<Section<uint32_t>*> sections;
  std::vector<Segment<uint32_t>*> segments;

  {
    auto seg_exe = new Segment<uint32_t>();
    seg_exe->InitExe(65536);
    segments.push_back(seg_exe);

    auto sec_null = new Section<uint32_t>();
    sec_null->InitNull();
    sections.push_back(sec_null);
    seg_exe->sections.push_back(sec_null);

    ASSERT(sec_text->data->size() > 0, "");
    sections.push_back(sec_text);
    seg_exe->sections.push_back(sec_text);
  }

  if (sec_rodata->data->size() > 0) {
    auto seg_ro = new Segment<uint32_t>();
    seg_ro->InitRO(65536);
    segments.push_back(seg_ro);

    sections.push_back(sec_rodata);
    seg_ro->sections.push_back(sec_rodata);
  }

  Segment<uint32_t>* seg_rw = nullptr;

  if (sec_data->data->size() + sec_bss->data->size() > 0) {
    seg_rw = new Segment<uint32_t>();
    seg_rw->InitRW(65536);
    segments.push_back(seg_rw);
  }

  if (sec_data->data->size() > 0) {
    sections.push_back(sec_data);
    seg_rw->sections.push_back(sec_data);
  }

  if (sec_bss->data->size() > 0) {
    sections.push_back(sec_bss);
    seg_rw->sections.push_back(sec_bss);
  }

  Section<uint32_t>* sec_symtab = nullptr;
  {
    auto seg_pseudo = new Segment<uint32_t>();
    seg_pseudo->InitPseudo();
    segments.push_back(seg_pseudo);
    auto sec_attr = new Section<uint32_t>();
    sec_attr->InitArmAttributes();
    sec_attr->SetData(
        new Chunk({ARM_ATTRIBUTES, sizeof(ARM_ATTRIBUTES)}, true));
    sections.push_back(sec_attr);
    seg_pseudo->sections.push_back(sec_attr);
    //
    if (create_sym_tab) {
      sec_symtab = new Section<uint32_t>();
      sec_symtab->InitSymTab(".symtab", sizeof(Elf_Sym<uint32_t>),
                             sections.size() + 1);
      sec_symtab->shdr.sh_info = symbols.size();
      std::string dummy(sizeof(Elf_Sym<uint32_t>) * symbols.size(), '\0');
      auto* sym_data = new Chunk(dummy, false);
      sec_symtab->SetData(sym_data);
      sections.push_back(sec_symtab);
      seg_pseudo->sections.push_back(sec_symtab);

      auto* names = new Chunk(padding_zero, false);
      for (auto& sym : symbols) {
        if (!sym->name.empty()) {
          sym->sym.st_name = names->size();
          names->AddData(sym->name);
          names->AddData(padding_zero);
        } else {
          sym->sym.st_name = 0;
        }
      }
      auto* sec_strtab = new Section<uint32_t>();
      sec_strtab->InitStrTab(".strtab");
      sec_strtab->SetData(names);
      sections.push_back(sec_strtab);
      seg_pseudo->sections.push_back(sec_strtab);
    }
    //
    auto sec_shstrtab = new Section<uint32_t>();
    sec_shstrtab->InitStrTab(".shstrtab");
    sections.push_back(sec_shstrtab);
    sec_shstrtab->SetData(MakeShStrTabContents(sections));
    seg_pseudo->sections.push_back(sec_shstrtab);
  }

  Executable<uint32_t> exe = MakeExecutableA32(0x20000, sections, segments);
  exe.UpdateVaddrsAndOffsets();
  for (auto& sym : symbols) {
    ASSERT(sym->sym.st_value != ~0, "undefined symbol " << sym->name);
    if (sym->section != nullptr) {
      ASSERT(sym->section->shdr.sh_addr != ~0,
             sym->name << "has bad sec " << *sym->section);
      sym->sym.st_value += sym->section->shdr.sh_addr;
      sym->sym.st_shndx = sym->section->index;
    }
  }

  for (auto& rel : relocations) {
    ApplyRelocation(rel);
  }

  if (create_sym_tab) {
    ASSERT(sec_symtab != nullptr, "");

    auto* sym_data = new Chunk("", false);
    for (auto& sym : symbols) {
      sym_data->AddData({(const char*)&sym->sym, sizeof(sym->sym)});
    }
    ASSERT(sec_symtab->shdr.sh_size == sym_data->size(), "");
    sec_symtab->SetData(sym_data);
  }

  auto* entry = global_symbol_map["_start"];
  ASSERT(entry != nullptr, "_start is not defined");
  exe.ehdr.e_entry = entry->sym.st_value;
  return exe;
}

std::ostream& operator<<(std::ostream& os, const Unit& s) {
  os << *s.sec_text << "\n"
     << *s.sec_rodata << "\n"
     << *s.sec_data << "\n"
     << *s.sec_bss << "\n";

  os << "String Table\n";
  for (const auto& sym : s.symbols) {
    os << *sym << "\n";
  }
  os << "Reloc Table\n";
  for (const auto& rel : s.relocations) {
    os << rel << "\n";
  }
  return os;
}

}  // namespace cwerg::a32

File: CpuA32/assembler.py
#!/usr/bin/python3

"""
This files contains ELF like abstraction to help build an assembler.
"""
from typing import List, Dict, Any

import CpuA32.opcode_tab as a32
import CpuA32.symbolic as dis
import Elf.elfhelper as elf
import Elf.enum_tab as elf_enum
from Elf import elf_unit

from Util import parse

NOP_BYTES = bytes([0x00, 0xF0, 0x20, 0xE3])


def DumpData(data: bytes, addr: int, syms: Dict[int, Any]) -> str:
    out = []
    first_address = True
    for n, b in enumerate(data):
        if first_address or (addr + n) % 8 == 0 or (addr + n) in syms:
            out.append(f"{addr:06x}")
            first_address = False
            name = syms.get(addr + n)
            if name:
                out[-1] += f" [{name}]"
        out[-1] += f" {b:02x}"
    return "\n".join(out)


def AddIns(unit: elf_unit.Unit, ins: a32.Ins):
    if ins.reloc_kind != elf_enum.RELOC_TYPE_ARM.NONE:
        sym = unit.FindOrAddSymbol(ins.reloc_symbol, ins.is_local_sym)
        unit.AddReloc(ins.reloc_kind, unit.sec_text, sym, ins.operands[ins.reloc_pos])
        # clear reloc info before proceeding
        ins.reloc_kind = elf_enum.RELOC_TYPE_ARM.NONE
        ins.operands[ins.reloc_pos] = 0
    unit.sec_text.AddData(a32.Assemble(ins).to_bytes(4, byteorder='little'))


def HandleOpcode(mnemonic, token: List[str], unit: elf_unit.Unit):
    AddIns(unit, dis.InsFromSymbolized(mnemonic, token))


def AddStartUpCode(unit: elf_unit.Unit):
    """Add code for `_start` wrapper which calls main(()

    When Linux transfers control to a new A32 program is does not follow any calling
    convention so we need this shim.
    The initial execution env looks like this:
    0(sp)			argc

    4(sp)			    argv[0] # argv start
    8(sp)               argv[1]
    ...
    (4*argc)(sp)        NULL    # argv sentinel

    (4*(argc+1))(sp)    envp[0] # envp start
    (4*(argc+2))(sp)    envp[1]
    ...
                        NULL    # envp sentinel
    """
    unit.FunStart("_start", 16, NOP_BYTES)
    for mnemonic, ops in [
        ("ldr_imm_add", "al r0 sp 0"),
        ("add_imm", "al r1 sp 4"),
        ("bl", "al lr expr:call:main"),
        ("movw", "al r7 1"),
        ("svc", "al 0"),
        ("ud2", "al")]:
        HandleOpcode(mnemonic, ops.split(), unit)
    unit.FunEnd()


class ParseError(Exception):
    pass


def UnitParse(fin, add_startup_code) -> elf_unit.Unit:
    unit = elf_unit.Unit()
    dir_handlers = {
        ".fun": lambda x, y: unit.FunStart(x, int(y, 0), NOP_BYTES),
        ".endfun": unit.FunEnd,
        ".mem": lambda x, y, z: unit.MemStart(x, int(y, 0), z, False),
        ".localmem": lambda x, y, z: unit.MemStart(x, int(y, 0), z, True),
        ".endmem": unit.MemEnd,
        ".data": lambda x, y: unit.AddData(int(x, 0),
                                           parse.QuotedEscapedStringToBytes(y)),
        ".addr.fun": lambda x, y: unit.AddFunAddr(elf_enum.RELOC_TYPE_ARM.ABS32, int(x, 0), y),
        ".addr.bbl": lambda x, y: unit.AddBblAddr(elf_enum.RELOC_TYPE_ARM.ABS32, int(x, 0), y),
        ".addr.mem": lambda x, y, z: unit.AddMemAddr(elf_enum.RELOC_TYPE_ARM.ABS32, int(x, 0), y, int(z, 0)),
        ".bbl": lambda x, y: unit.AddLabel(x, int(y, 0), NOP_BYTES),
    }
    for line_num, line in enumerate(fin):
        token = parse.ParseLine(line)
        # print(line)
        if not token:
            continue
        mnemonic = token.pop(0)

        if mnemonic.startswith("#"):
            continue
        if mnemonic.startswith("."):
            handler = dir_handlers.get(mnemonic)
            assert handler is not None
            handler(*token)
            continue

        try:
            HandleOpcode(mnemonic, token, unit)
        except Exception as err:
            raise ParseError(
                f"UnitParseFromAsm error in line {line_num}:\n{line}\n{token}\n{err}")
    unit.AddLinkerDefs()
    if add_startup_code:
        AddStartUpCode(unit)
    return unit


_OPCODE_MOVW: a32.Opcode = a32.Opcode.name_to_opcode["movw"]
_OPCODE_MOVT: a32.Opcode = a32.Opcode.name_to_opcode["movt"]
_OPCODE_B: a32.Opcode = a32.Opcode.name_to_opcode["b"]
_OPCODE_BL: a32.Opcode = a32.Opcode.name_to_opcode["bl"]


def _branch_offset(rel: elf.Reloc, sym_val: int) -> int:
    return (sym_val - rel.section.sh_addr - rel.r_offset - 8) >> 2


def _ApplyRelocation(rel: elf.Reloc):
    sec_data = rel.section.data
    sym_val = rel.symbol.st_value + rel.r_addend
    assert rel.r_offset + 4 <= len(sec_data)
    old_data = int.from_bytes(sec_data[rel.r_offset:rel.r_offset + 4], "little")

    if rel.r_type == elf_enum.RELOC_TYPE_ARM.MOVW_ABS_NC.value:
        new_data = a32.Patch(old_data, _OPCODE_MOVW, 2, sym_val & 0xffff)
    elif rel.r_type == elf_enum.RELOC_TYPE_ARM.MOVT_ABS.value:
        new_data = a32.Patch(old_data, _OPCODE_MOVT, 2, (sym_val >> 16) & 0xffff)
    elif rel.r_type == elf_enum.RELOC_TYPE_ARM.JUMP24.value:
        new_data = a32.Patch(old_data, _OPCODE_B, 1, _branch_offset(rel, sym_val))
    elif rel.r_type == elf_enum.RELOC_TYPE_ARM.CALL.value:
        new_data = a32.Patch(old_data, _OPCODE_BL, 2, _branch_offset(rel, sym_val))
    elif rel.r_type == elf_enum.RELOC_TYPE_ARM.ABS32.value:
        new_data = sym_val
    else:
        assert False, f"unknown kind reloc {rel}"

    sec_data[rel.r_offset:rel.r_offset + 4] = new_data.to_bytes(4, "little")
    print(f"PATCH INS {rel.r_type} {rel.r_offset:x} {sym_val:x} {old_data:x} {new_data:x} {rel.symbol.name}")


def Assemble(unit: elf_unit.Unit, create_sym_tab: bool) -> elf.Executable:
    sections = []
    segments = []

    seg_exe = elf.Segment.MakeExeSegment(65536)
    segments.append(seg_exe)

    sec_null = elf.Section.MakeSectionNull()
    sections.append(sec_null)
    seg_exe.sections.append(sec_null)
    #
    sec_text = unit.sec_text
    assert len(sec_text.data) > 0
    sections.append(sec_text)
    seg_exe.sections.append(sec_text)

    sec_rodata = unit.sec_rodata
    if len(sec_rodata.data) > 0:
        seg_ro = elf.Segment.MakeROSegment(65536)
        segments.append(seg_ro)
        #
        sections.append(sec_rodata)
        seg_ro.sections.append(sec_rodata)

    if len(unit.sec_data.data) + len(unit.sec_bss.data) > 0:
        seg_rw = elf.Segment.MakeRWSegment(65536)
        segments.append(seg_rw)

    sec_data = unit.sec_data
    if len(sec_data.data) > 0:
        sections.append(sec_data)
        seg_rw.sections.append(sec_data)

    sec_bss = unit.sec_bss
    if len(sec_bss.data) > 0:
        sections.append(sec_bss)
        seg_rw.sections.append(sec_bss)

    seg_pseudo = elf.Segment.MakePseudoSegment()
    segments.append(seg_pseudo)
    #
    sec_attr = elf.Section.MakeSectionArmAttributes()
    sections.append(sec_attr)
    sec_attr.SetData(elf.ARM_ATTRIBUTES)
    seg_pseudo.sections.append(sec_attr)

    if create_sym_tab:
        # we do not create the content here since we cannot really do this until
        # the section addresses are finalized
        which = elf_enum.EI_CLASS.X_32
        sec_symtab = elf.Section.MakeSectionSymTab(".symtab", which, len(sections) + 1)
        sections.append(sec_symtab)
        sec_symtab.SetData(bytearray(len(unit.symbols) * elf.Symbol.SIZE[which]))
        seg_pseudo.sections.append(sec_symtab)
        # TODO: this is not quite right
        sec_symtab.sh_info = len(unit.symbols)

        sym_names = bytearray()
        sym_names += b"\0"
        for sym in unit.symbols:
            if sym.name:
                sym.st_name = len(sym_names)
                sym_names += (bytes(sym.name, "utf-8") + b"\0")
            else:
                sym.st_name = 0
        sec_symstrtab = elf.Section.MakeSectionStrTab(".strtab")
        sections.append(sec_symstrtab)
        sec_symstrtab.SetData(sym_names)
        seg_pseudo.sections.append(sec_symstrtab)

    sec_shstrtab = elf.Section.MakeSectionStrTab(".shstrtab")
    sections.append(sec_shstrtab)
    sec_shstrtab.SetData(elf.MakeSecStrTabContents(sections))
    seg_pseudo.sections.append(sec_shstrtab)

    exe = elf.Executable.MakeExecutableA32(0x20000, sections, segments)
    exe.update_vaddrs_and_offset()

    if False:
        for sym in unit.symbols:
            print("@@@SYMS", sym)
        for rel in unit.relocations:
            print("@@@REL", rel)

    for sym in unit.symbols:
        if sym.section:
            assert sym.section.sh_addr > 0
            assert sym.st_value != elf.TO_BE_FILLED_IN_LATER
            sym.st_value += sym.section.sh_addr
            sym.st_shndx = sym.section.index

    for rel in unit.relocations:
        _ApplyRelocation(rel)

    if create_sym_tab:
        # we only put dummiess in the symtable above - do it for real now
        sec_symtab.data = bytearray()
        for sym in unit.symbols:
            sec_symtab.AddData(sym.pack(which))

    sym_entry = unit.global_symbol_map["_start"]
    assert sym_entry and not sym_entry.is_undefined()
    entry_addr = sym_entry.st_value
    print(f"PATCH ENTRY: {entry_addr:x}")
    exe.ehdr.e_entry = entry_addr
    return exe
